### PR TITLE
WIP: [luci/service] Support BMM dynamic shape inferece

### PR DESCRIPTION
**Pull request by zetwhite**
This PR supports dynamic shpae inference ofr BatchMatMul Op.

ONE-DCO-1.0-Signed-off-by: SeungHui Youn <sseung.youn@samsung.com>

### CODE REVIEW

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp [hunk 1]
```diff
@@ -30,4 +52,56 @@ luci::CircleNode *CloneNodeLet<CN::ABC>::visit(const luci::CircleBatchMatMul *no
...
+  const auto y_shape = sinf::circle_shape(y);
+
+  uint32_t x_rank = x_shape.rank();
+  uint32_t y_rank = y_shape.rank();
+  assert(x_rank >= 2 && y_rank >= 2);
+
+  uint32_t max_rank = x_rank > y_rank ? x_rank : y_rank;
+  loco::TensorShape output_shape;
+  output_shape.rank(max_rank);
+
```
Comment (zetwhite) [inline compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:72]: If this kind of comment is helpful to understand, please let me know. 
I'll append it to the code. 

```c++
  // BatchMatMul shape inference rule works with two-part
  // 
  // 1) Batch dimensions part
  //    - Batch dimensions corresponds the to input_shape[:-2]
  //    - General broadcast rules are used to infer the shape of the output batch
  // 
  // 2) Contracting dimensions part
  //    - Contracting dimensions corresponds to the input_shape[-2:]
  //    - General matrix multiplication shape inference applied for this part, 
  //      which means '(x_lhs, x_rhs) x (y_lhs, y_rhs) => (x_lhs, y_rhs)'
```

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp [hunk 2]
```diff
@@ -30,4 +52,56 @@ luci::CircleNode *CloneNodeLet<CN::ABC>::visit(const luci::CircleBatchMatMul *no
...
+    const auto o_batch_rank = o_batch_dims.rank();
+    for (uint i = 0u; i < o_batch_rank; ++i)
+    {
+      output_shape.dim(i) = o_batch_dims.dim(i);
+    }
+  }
+
+  // shape inference in contracting dimensions
+  auto adj_x = node->adj_x();
+  auto adj_y = node->adj_y();
```
Comment (zetwhite) [inline compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:90]:
```suggestion
  const auto adj_x = node->adj_x();
  const auto adj_y = node->adj_y();
```

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp [hunk 3]
```diff
@@ -30,4 +52,56 @@ luci::CircleNode *CloneNodeLet<CN::ABC>::visit(const luci::CircleBatchMatMul *no
...
+{
+  const auto x = loco::must_cast<CircleNode *>(node->x());
+  const auto y = loco::must_cast<CircleNode *>(node->y());
+
+  const auto x_shape = sinf::circle_shape(x);
+  const auto y_shape = sinf::circle_shape(y);
+
+  uint32_t x_rank = x_shape.rank();
+  uint32_t y_rank = y_shape.rank();
+  assert(x_rank >= 2 && y_rank >= 2);
```
Comment (seanshpark) [inline compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:67]: just to be sure, `x_rank` and `y_rank` must be `>= 2`
and if not this model is invalid ?

Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:67]: > and if not this model is invalid ? 

Yes. I think we can consider 'input with `rank < 2` BatchMatMul is invalid'.  

Because both TFLite and Pytorch don't allow `rank < 2` inputs for BatchMatMul :  

* TFLite `ops::BatchMatMul` (https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/batch-mat-mul) allows 2-D or higher inputs 
  > Args: 
   > x: 2-D or higher with shape [..., r_x, c_x]. 
   > y: 2-D or higher with shape [..., r_y, c_y]. 

* pytorch `torch.bmm` (https://pytorch.org/docs/stable/generated/torch.bmm.html) allows only D inputs (>= 2D) 
  > input and mat2 must be 3-D tensors each containing the same number of matrices.

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp [hunk 4]
```diff
@@ -30,4 +52,56 @@ luci::CircleNode *CloneNodeLet<CN::ABC>::visit(const luci::CircleBatchMatMul *no
...
+
+  uint32_t max_rank = x_rank > y_rank ? x_rank : y_rank;
+  loco::TensorShape output_shape;
+  output_shape.rank(max_rank);
+
+  // broadcast in the batch dimensions
+  if (x_rank > 2 || y_rank > 2)
```
Comment (ys44kim) [inline compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:74]: if x_rank = 3, y_rank = 2, the condition also works,
In this case, y_batch_dims rank is 0, is this also allowed?
i want to know how it works because there is no test case.

File: compiler/luci/service/src/CircleShapeInferenceRule.cpp [hunk 5]
```diff
@@ -463,52 +463,6 @@ template <class Conv2DType> OutputSize infer_conv2d_type(const Conv2DType *node)
...
 }
 
-// BatchMatMulV2 supports broadcasting in the batch dimensions(BatchMatMul doesn't)
-// TODO Distinguish BatchMatMul and BatchMatMulV2
-loco::NodeShape infer_batchmatmul_shape(const loco::TensorShape &x_shape,
-                                        const loco::TensorShape &y_shape, bool adj_x, bool adj_y)
-{
-  uint32_t x_rank = x_shape.rank();
-  uint32_t y_rank = y_shape.rank();
-  assert(x_rank >= 2 && y_rank >= 2);
```
Comment (zetwhite) [inline compiler/luci/service/src/CircleShapeInferenceRule.cpp:473]: (note) Original shape inference also allows inputs with rank >=2

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp (hunk 3 above)
Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:67]: Also, original shape inference uses assertion to check `rank >=2`. 
 https://github.com/Samsung/ONE/pull/13759/files#r1732652062

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp (hunk 4 above)
Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:74]: > In this case, y_batch_dims rank is 0, is this also allowed? 

Yes.  It is allowed. 

> i want to know how it works because there is no test case. 

I make an example for the case. 
<sub> I intentionally use ? instead of 0, because 0 means unknown in shape, here </sub> 
```
x: (?, 4, 3) ----  BatchMatMul --- out: (?, 4, 5) 
		  / 	
y: (3, 5)    ----/ 

```

Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:74]: Ah, I a bit misunderstood your questions :smiling_face_with_tear: 

I put another working example of 2D - 3D batch matmul.  
```
x: (2, 4, 3) ----  BatchMatMul --- out: (2, 4, 5) 
		          / 	
y: (3, 5)    ----/ 
```

**Comment by shs-park** [comment]
I made some test cases with `tensorflow` python code.
I could get exceptions in `Test 4` and `Test 6` cases.

Please note when you make test cases.

```python
import tensorflow as tf

# Example 1: Simple 2D Matrices Batch
A = tf.constant([[[1, 2, 3], [4, 5, 6]],
                 [[7, 8, 9], [10, 11, 12]]], dtype=tf.float32)  # Shape: (2, 2, 3)

B = tf.constant([[[1, 4], [2, 5], [3, 6]],
                 [[7, 10], [8, 11], [9, 12]]], dtype=tf.float32)  # Shape: (2, 3, 2)

# Perform batch matrix multiplication using tf.matmul
result = tf.matmul(A, B)
print("Example 1 Result (Shape: {}):\n{}".format(result.shape, result.numpy()))

# Example 2: Broadcasting in Batch Dimensions
A = tf.constant([[[1, 2, 3], [4, 5, 6]],
                 [[7, 8, 9], [10, 11, 12]],
                 [[13, 14, 15], [16, 17, 18]]], dtype=tf.float32)  # Shape: (3, 2, 3)

B = tf.constant([[[1, 4], [2, 5], [3, 6]]], dtype=tf.float32)  # Shape: (1, 3, 2)

# Perform batch matrix multiplication with broadcasting using tf.matmul
result = tf.matmul(A, B)
print("Example 2 Result (Shape: {}):\n{}".format(result.shape, result.numpy()))

# Example 3: Edge case with zero-dimension batch
A = tf.random.normal([0, 2, 3])  # Shape: (0, 2, 3)
B = tf.random.normal([0, 3, 2])  # Shape: (0, 3, 2)

# Perform batch matrix multiplication using tf.matmul
result = tf.matmul(A, B)
print("Example 3 Result (Shape: {}):\n{}".format(result.shape, result.numpy()))

# Example 3-2: Edge case with zero-dimension batch
A = tf.random.normal([3, 2, 3])  # Shape: (3, 2, 3)
B = tf.random.normal([3, 2])  # Shape: (3, 2)

# Perform batch matrix multiplication using tf.matmul
result = tf.matmul(A, B)
print("Example 3-2 Result (Shape: {}):\n{}".format(result.shape, result.numpy()))

# Example 4: Edge case with zero-dimension batch
A = tf.random.normal([2, 2, 3])  # Shape: (2, 2, 3)
B = tf.random.normal([0, 3, 2])  # Shape: (0, 3, 2)

# Perform batch matrix multiplication using tf.matmul
#result = tf.matmul(A, B)
print("Example 4 Result EXCEPTION")

# Example 5: Edge case with rank-2 batch
A = tf.random.normal([2, 3])  # Shape: (2, 3)
B = tf.random.normal([3, 2])  # Shape: (3, 2)

# Perform batch matrix multiplication using tf.matmul
result = tf.matmul(A, B)
print("Example 5 Result (Shape: {}):\n{}".format(result.shape, result.numpy()))

# Example 6: Edge case with rank-2 and rank-1 atch
A = tf.random.normal([2, 3])  # Shape: (2, 3)
B = tf.random.normal([3])  # Shape: (3)

# Perform batch matrix multiplication using tf.matmul
#result = tf.matmul(A, B)
print("Example 6 Result EXCEPTION")
```

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp [hunk 6]
```diff
@@ -35,3 +37,290 @@ TEST(CloneNodeTest, clone_BatchMatMul)
...
+  bmm.y(&input_y);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  // (2, 40, 40) x (1, 30, 70)
+  //         ^         ^
+  // => error, matmul hidden dim should be same
+  ASSERT_ANY_THROW(shape_inf_rule.infer(&bmm, shape));
+}
```
Comment (shs-park) [inline compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp:326]: How about adding zero-dim cases but it is static. (actual 0-dim)
For example in python code,

```python
# Example 3: Edge case with zero-dimension batch
A = tf.random.normal([0, 2, 3])  # Shape: (0, 2, 3)
B = tf.random.normal([0, 3, 2])  # Shape: (0, 3, 2)
```
-> The shape of result should be (0, 2, 2)

and,

```python
# Example 4: Edge case with zero-dimension batch
A = tf.random.normal([2, 2, 3])  # Shape: (2, 2, 3)
B = tf.random.normal([0, 3, 2])  # Shape: (0, 3, 2)
```
-> The should generate an exception

**Review by shs-park — COMMENTED** [review]
I left some comments, PTAL
=)

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp (hunk 3 above)
Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:67]: I leave an additional comment about this - https://github.com/Samsung/ONE/pull/13759#discussion_r1732649791. 

I thought that the circle model is imported from TensorFlow or PyTorch, and if both frameworks do not allow inputs under 2D, such a circle model would not exist. So I thought we don't need to consider such cases. 

------------------------------------------

By the way, using an exception is also fine with me. If others think using an exception is better for safety, I'll fix it.

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp (hunk 6 above)
Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp:326]: > How about adding zero-dim cases but it is static. (actual 0-dim) 

I guess if tensor has a 0 dimension, it means the tensor is empty. 
I thought that graph with empty tensor is invalid, and I didn't handle this case in shape inference. 

I'd better add an `assertion` or `exception` for this case. (also negative test case if possible). 
Thanks for catching.

File: compiler/luci/service/src/CircleShapeInferenceRule.cpp (hunk 5 above)
Comment (seanshpark) [reply compiler/luci/service/src/CircleShapeInferenceRule.cpp:473]: well, original code may have problems too =_=;;;

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp (hunk 3 above)
Comment (jinevening) [reply compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:67]: +1 for exception. TF, Torch or our converting tools (circle-exir, onnx2circle) can have a bug. It would be better to explicitly check invalid inputs.

**Comment by jinevening** [comment]
LGTM with minor comment https://github.com/Samsung/ONE/pull/13759/files#r1733773189

**Review by seanshpark — APPROVED** [review]
LGTM

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp [hunk 7]
```diff
@@ -30,4 +76,72 @@ luci::CircleNode *CloneNodeLet<CN::ABC>::visit(const luci::CircleBatchMatMul *no
...
+  const auto y_shape = sinf::circle_shape(y);
+
+  uint32_t x_rank = x_shape.rank();
+  uint32_t y_rank = y_shape.rank();
+
+  // throw internal exception if condition not met
+  throw_unless(x_rank >= 2, "x_rank shoud be >= 2");
+  throw_unless(y_rank >= 2, "y_rank shoud be >= 2");
+  throw_unless((not contain_zero(x_shape)), "x_shape should NOT have 0");
+  throw_unless((not contain_zero(y_shape)), "y_shape should NOT have 0");
```
Comment (zetwhite) [inline compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:96]: to resolve https://github.com/Samsung/ONE/pull/13759/files#r1732922382

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp [hunk 8]
```diff
@@ -35,3 +36,369 @@ TEST(CloneNodeTest, clone_BatchMatMul)
...
+  bmm.y(&input_y);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  // (0, 2) x (2, 4)
+  //  ^
+  // => error, x should not be empty
+  ASSERT_ANY_THROW(shape_inf_rule.infer(&bmm, shape));
+}
```
Comment (zetwhite) [inline compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp:404]: (newly added) to handle empty input

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp (hunk 6 above)
Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp:356]: I updated the logic - https://github.com/Samsung/ONE/pull/13759/files#r1733812036, and test case - https://github.com/Samsung/ONE/pull/13759/files#r1733813949

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp [hunk 9]
```diff
@@ -35,3 +36,369 @@ TEST(CloneNodeTest, clone_BatchMatMul)
...
+  bmm.x(&input_x);
+  bmm.y(&input_y);
+
+  loco::TensorShape shape;
+  luci::sinf::Rule shape_inf_rule;
+
+  ASSERT_TRUE(shape_inf_rule.infer(&bmm, shape));
+
+  // (5, 4, 3) x (3, 8) -> (5, 3, 8)
+  // output shape should be (5, 3, 8)
```
Comment (zetwhite) [inline compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp:92]:
```suggestion
  // output shape should be (5, 4, 8)
```

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp [hunk 10]
```diff
@@ -35,3 +36,369 @@ TEST(CloneNodeTest, clone_BatchMatMul)
...
+  // (5, 4, 3) x (3, 8) -> (5, 3, 8)
+  // output shape should be (5, 4, 8)
+  ASSERT_EQ(3, shape.rank());
+  ASSERT_TRUE(shape.dim(0).known());
+  ASSERT_TRUE(shape.dim(1).known());
+  ASSERT_TRUE(shape.dim(2).known());
+  ASSERT_EQ(5, shape.dim(0).value());
+  ASSERT_EQ(4, shape.dim(1).value());
+  ASSERT_EQ(8, shape.dim(2).value());
+}
```
Comment (zetwhite) [inline compiler/luci/service/src/Nodes/CircleBatchMatMul.test.cpp:100]: @ys44kim I added the TC for the case you asked about. :)

**Review by seanshpark — APPROVED** [review]
LGTM

**Review by shs-park — APPROVED** [review]
Thank you, LGTM!
=)

**Review by jinevening — APPROVED** [review]
LGTM

**Comment by jinevening** [comment]
Ping @ys44kim

**Review by ys44kim — APPROVED** [review]
LGTM

File: compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp [hunk 11]
```diff
@@ -30,4 +76,72 @@ luci::CircleNode *CloneNodeLet<CN::ABC>::visit(const luci::CircleBatchMatMul *no
   return cloned;
 }
 
+// BatchMatMulV2 supports broadcasting in the batch dimensions(BatchMatMul doesn't)
+// TODO Distinguish BatchMatMul and BatchMatMulV2
+loco::TensorShape sinf::Algorithm::visit(const luci::CircleBatchMatMul *node)
```
Comment (seanshpark) [inline compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:81]: @zetwhite , please check https://github.com/Samsung/ONE/pull/13780#pullrequestreview-2265214545

can you apply here too? or as getting approvals takes time and you've go all ACKs, 
apply in another PR?

Comment (seanshpark) [reply compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:81]: if go with another PR, you can apply changes to all Ops (I think there aren't too many)

Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:81]: I'd like to merge this PR as is. 
If fine, I'll make another PR to fix it.

Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp:81]: > you can apply changes to all Ops (I think there aren't too many) 

Oh, I see. I apply it at once.

**Comment by seanshpark** [comment]
ah... commit title had typo ... -_-;
